rfcs/20200205-standalone-keras-repository.md

```diff
@@ -0,0 +1,351 @@
+and the other as a PR to the Keras repo. Here are some common scenarios:
+
+1. Adding a new feature to TensorFlow, and having Keras rely on it. Note that the 
+TF change needs to be submitted first, and the Keras PR needs to wait for the new TF
+nightly to become available on PyPI.
+
+Also note that any rollback of the TF PR will cause Keras to break, the
+rollback sequence should be PR 33333 and then PR 22222 (see example below).
+The Google-internal test for TF should catch the error if the rollback sequence
+is not correct.
+
+```python
+# Existing scenario.
+# PR 11111 (2 files updated)
+# +++ tensorflow/python/ops/array_ops.py
+def some_new_function(inputs):
+   ...
+
+# +++ tensorflow/python/keras/layers/core.py
+
+class new_layer(Layer):
+
+  def call(inputs):
+     array_ops.some_new_function(inputs)
+     ...
+```
+
+```python
+# New scenario.
+# PR 22222 (1 file updated)
+# +++ tensorflow/python/ops/array_ops.py
+@tf.export('some_new_function')
+def some_new_function(inputs):
+   ...
+
+==================================
+# PR 33333 (1 file updated)
+# +++ tensorflow/python/keras/layers/core.py
+
+class new_layer(Layer):
+
+  def call(inputs):
+     tf.some_new_function(inputs)
+     ...
+```
+
+2. Changing the behavior of an existing TF API.
+
+Note that the PR 22222 needs to be submitted with both the new and old
+function since Google internal CI is still testing from HEAD.
+The previous function can be 
+deleted after PR 33333 is submitted. Also note that this issue is caused by 
+Keras not using exclusively public TF API, but relying on TF implementation details.
+Moving towards only using public APIs should reduce the likelihood of this kind of issue.
+
+```python
+# Existing scenario.
+# PR 11111 (2 files updated)
+# tensorflow/python/ops/array_ops.py
+<<<
+def existing_function(inputs):
+    ...
+>>>
+def new_function(inputs, knob1=False, knob2=1):
+    ...
+# tensorflow/python/keras/layers/core.py
+
+class existing_layer(Layer):
+
+  def call(inputs):
+<<<
+    array_ops.existing_function(inputs)
+>>>
+    array_ops.new_function(
+        inputs, 
+        knob1=True,
+        knob2=3)
+```
+
+```python
+# New scenario.
+# PR 22222 (1 file updated)
+# tensorflow/python/ops/array_ops.py
+<<<
+def existing_function(inputs):
+   ...
+>>>
+def existing_function(inputs):
+  return new_function(
+    inputs, 
+    knob1=False,
+    knob2=1)
+
+def new_function(inputs, knob1, knob2=1):
+    ...
+
+==================================
+# PR 33333 (1 file updated)
+# tensorflow/python/keras/layers/core.py
+class existing_layer(Layer):
+
+  def call(inputs):
+<<<
+    array_ops.existing_function(inputs)
+     ...
+>>>
+    array_ops.new_function(
+        inputs, 
+        knob1=True,
+        knob2=3)
+```
+
+
+### Performance Implications
+
+There may be some performance implications as we move towards only using
+public TF APIs. We need to maintain a benchmark to ensure that there
+is no performance regression.
+
+### Dependencies
+
```

Comment:
Can you explain how namespacing will work without a circular dep? 
Keras depends on TF, will Keras continue to be under tf namespace as `tf.keras`?

Reply:
Good question, will we be able to do `import keras`?

Reply:
If we do the same thing as we did for keras-prepro and keras-app, it should work right?

Reply:
@Dref360 can you detail a bit? I'm sure there are people not familiar with your work (including me 😄 )

Reply:
Anna's design covers more details about the PIP package structure. https://github.com/tensorflow/community/pull/182. Currently TF package is an umbrella package, and the tensorflow_core is the package that contains all the real code. The new Keras package will depend on tensorflow_core, and TF umbrella package will depend on Keras.

This the same approach used for tensorflow_estimator, which relies on Tensorflow ops, but still exposed under tf.estimator.

Reply:
@gabrieldemarmiesse, yes. tf.keras symbol will still work, and so as keras. Under the hood, tf.keras will load the python symbol from keras.

Reply:
It would be nice to have 
```python
import keras
```

instead of 
```
import tensorflow_keras as keras
```
But if you prefer @qlzh727 we can move the discussion to #182

Reply:
Yes, I had offline discussion with Anna before, and we would prefer keras instead of Keras. I think when Anna wrote her design, she was not sure about what package we will use, and she assumed we will use the similar name as tensorflow_estimator, which is why she put tensorflow_keas in her design. @annarev for #182.

Reply:
Very nice, thanks! Some users won't even see the transition when upgrading their keras multi-backend ;)

Reply:
what @qlzh727 is right, I didn't mean to imply `tensorflow_keras` as the package name to use in my doc, just used it similarly with `tensorflow_estimator`.

Reply:
Can we add it somewhere in the document to set it in stone? :)

Reply:
This has been adressed by @annarev in https://github.com/tensorflow/community/pull/182#discussion_r377591797 @qlzh727 I believe you can minimize this thread.